diff --git a/e2e/bug10449b/src/Query/ExampleQueryHandler.php b/e2e/bug10449b/src/Query/ExampleQueryHandler.php
index e9bc1d59f..f5d1c07bf 100644
--- a/e2e/bug10449b/src/Query/ExampleQueryHandler.php
+++ b/e2e/bug10449b/src/Query/ExampleQueryHandler.php
@@ -7,13 +7,13 @@ namespace App\Query;
 use App\Bus\QueryHandlerInterface;
 
 /**
- * @phpstan-type Return string
+ * @phpstan-type Return int
  */
 final class ExampleQueryHandler implements QueryHandlerInterface
 {
     /** @return Return */
     public function __invoke(ExampleQuery $exampleQuery)
     {
-        return '1';
+        return 1;
     }
 }
\ No newline at end of file
